test/NuGet.Core.FuncTests/NuGet.Protocol.FuncTest/AuthenticationHandlerTests.cs

```diff
@@ -0,0 +1,215 @@
+        }
+
+        [Fact]
+        public async Task OnlyOneRequestWithoutAuthorizationHeader()
+        {
+            // Arrange
+            var portReserver = new PortReserver();
+            await portReserver.ExecuteAsync(async (port, cancellationToken) =>
+            {
+                var server = new RequestCollectingServer(_output);
+                server.Start(port);
+                try
+                {
+                    var packageSource = new PackageSource(server.BaseUrl + "v3/index.json", "test");
+                    packageSource.Credentials = new PackageSourceCredential(packageSource.Name, "user", "pass", true, "basic");
+                    var repository = Repository.Factory.GetCoreV3(packageSource);
+
+                    var httpSourceResource = await repository.GetResourceAsync<HttpSourceResource>();
+                    var source = httpSourceResource.HttpSource;
+
+                    using (var sourceCacheContext = new SourceCacheContext()
+                    {
+                        DirectDownload = true,
+                        NoCache = true,
+                    })
+                    {
+                        Mock<ILogger> logger = new();
+                        HttpSourceCacheContext httpSourceCacheContext = HttpSourceCacheContext.Create(sourceCacheContext, isFirstAttempt: true);
+
+                        // Act
+                        var request = new HttpSourceCachedRequest(server.BaseUrl + "v3/index.json", "1", httpSourceCacheContext);
+                        _ = await source.GetAsync(request, ProcessResponse, logger.Object, cancellationToken);
+
+                        request = new HttpSourceCachedRequest(server.BaseUrl + "v3/flatcontainer/packageid/1.2.3/index.json", "2", httpSourceCacheContext);
+                        _ = await source.GetAsync(request, ProcessResponse, logger.Object, cancellationToken);
+                    }
+                }
+                finally
+                {
+                    server.Stop();
+                }
+
+                // Assert
+                Assert.Equal(
+                    1,
+                    server.Requests.Count(RequestWithoutAuthorizationHeader));
+                Assert.Equal(
+                    2,
+                    server.Requests.Select(r => r.RawUrl).Distinct().Count());
+
+                // ExecuteAsync returns Task<T>, so need to return something to give it a <T>.
+                return (object?)null;
+            },
+            CancellationToken.None);
+
+            static bool RequestWithoutAuthorizationHeader(HttpListenerRequest request)
+            {
+                string? value = request.Headers["Authorization"];
+                return string.IsNullOrEmpty(value);
+            }
+
+            static Task<string> ProcessResponse(HttpSourceResult result)
+            {
+                return Task.FromResult(string.Empty);
+            }
+        }
+
+        private class RequestCollectingServer
```

Comment:
The class appears generic enough to put alongside similar ones in our Test.Utility/TestServer folder.